### ISSUE TITLE
Unexpected behaviour when creating a conflict

### ISSUE THREAD

**Issue by canton7** [syncthing/syncthing#3992]
I came across some unexpected behaviour when testing conflicts - repro below.

1. Two devices, "Local" and "Remote". Remote is v0.14.23, Linux (64 bit), Local is v0.14.23+14-gfb6d453, Windows (64 bit). Folder shared ("Syncbox" below). A file is created, and is in sync between the two ("test/testfile.txt" below). Both have high rescan intervals.
2. Modify the file on both, with different contents.
3. Click "Rescan" on Remote.

One of two things tends to happen - both of which look a bit odd:

## Option 1

There's a flurry of activity. `-verbose` on the Local device shows this:

```[LOCAL] 12:28:33 VERBOSE: Device REMOTE sent an index update for "SyncBox" with 1 items
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now syncing
[LOCAL] 12:28:33 INFO: Folder "SyncBox" isn't making progress. Pausing puller for 1m0s.
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Summary for folder "SyncBox" is map[localDeleted:20 needBytes:2 needDirectories:0 inSyncFiles:53 globalFiles:54 globalDeleted:20 sequence:660 localDirectories:10 globalBytes:11787503 version:660 localBytes:11787502 needFiles:1 needDeletes:0 state:scanning localSymlinks:0 globalDirectories:10 globalSymlinks:0 needSymlinks:0 localFiles:54 inSyncBytes:11787501]
[LOCAL] 12:28:33 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 100%
[LOCAL] 12:28:33 VERBOSE: Local change detected in folder "": modified file C:\Documents\SyncBox\test\testfile.txt
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 12:28:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:28:35 VERBOSE: Summary for folder "SyncBox" is map[needFiles:1 inSyncBytes:11787501 globalSymlinks:0 localSymlinks:0 needSymlinks:0 localBytes:11787502 needBytes:2 state:idle localFiles:54 needDeletes:0 needDirectories:0 globalFiles:54 globalDirectories:10 version:661 inSyncFiles:53 sequence:661 globalBytes:11787503 localDirectories:10 localDeleted:20 globalDeleted:20]
[LOCAL] 12:28:35 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 100%
[LOCAL] 12:28:44 INFO: Joined relay relay://78.46.205.180:22067
[LOCAL] 12:28:53 VERBOSE: Listen address dynamic+https://relays.syncthing.net/endpoint resolution has changed: lan addresses: [relay://78.46.205.180:22067/?id=455XXKT-FPQYVLX-DSPTWGQ-OMIGZYH-EYHRVQH-UB6G2RR-YJ375FH-BR5XYAY&pingInterval=1m0s&networkTimeout=2m0s&sessionLimitBps=1073741824&globalLimitBps=4294967296&statusAddr=:22070&providedBy=ae] wan addresses: [relay://78.46.205.180:22067/?id=455XXKT-FPQYVLX-DSPTWGQ-OMIGZYH-EYHRVQH-UB6G2RR-YJ375FH-BR5XYAY&pingInterval=1m0s&networkTimeout=2m0s&sessionLimitBps=1073741824&globalLimitBps=4294967296&statusAddr=:22070&providedBy=ae]
```

(Note the flicking between idle and scanning).

Remote shows Local as "Syncing (100%)". Local shows the folder as "Out of Sync", with "1 failed item". Clicking the "1 failed" link however gives this:

![thing](https://cloud.githubusercontent.com/assets/568104/22974600/03f0ed80-f37b-11e6-9d21-62ab0c518426.png)

After a minute, Local recovers itself and creates the conflict file:

```
[LOCAL] 12:29:33 VERBOSE: Folder "SyncBox" is now syncing
[LOCAL] 12:29:33 VERBOSE: Started syncing "SyncBox" / "test\\testfile.txt" (update file)
[LOCAL] 12:29:33 INFO: Attempted to rename \\?\C:\Documents\SyncBox\test\testfile.txt to \\?\C:\Documents\SyncBox\test\testfile.sync-conflict-20170215-122933.txt got <nil>
[LOCAL] 12:29:33 VERBOSE: Finished syncing "SyncBox" / "test\\testfile.txt" (update file): Success
[LOCAL] 12:29:33 VERBOSE: Remote change detected in folder "": modified file C:\Documents\SyncBox\test\testfile.txt
[LOCAL] 12:29:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:29:33 VERBOSE: Summary for folder "SyncBox" is map[globalDirectories:10 needDeletes:0 globalFiles:54 globalBytes:11787503 localDeleted:20 globalSymlinks:0 inSyncFiles:54 sequence:662 needFiles:0 inSyncBytes:11787503 localSymlinks:0 needSymlinks:0 state:idle localBytes:11787503 globalDeleted:20 needBytes:0 localFiles:54 localDirectories:10 needDirectories:0 version:662]
[LOCAL] 12:29:33 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 99.99998303287812%
[LOCAL] 12:29:33 VERBOSE: Device REMOTE sent an index update for "SyncBox" with 1 items
[LOCAL] 12:29:33 VERBOSE: Folder "SyncBox" is now syncing
[LOCAL] 12:29:33 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 12:29:33 VERBOSE: Summary for folder "SyncBox" is map[version:663 needDeletes:0 needSymlinks:0 inSyncFiles:54 localDeleted:20 needFiles:0 globalBytes:11787503 needBytes:0 localDirectories:10 globalDeleted:20 localBytes:11787503 sequence:663 needDirectories:0 globalSymlinks:0 localSymlinks:0 localFiles:54 state:idle globalDirectories:10 inSyncBytes:11787503 globalFiles:54]
[LOCAL] 12:29:33 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 100%
```

However the conflict file is not synced back to Remote until I manually rescan Local.

## Option 2

The conflict file is created on Remote (which is odd - I would have thought it would be created on Local), and then isn't synced to Local. Both devices appear to be in sync.

Here's the verbose log from Local again:

```
[LOCAL] 13:20:17 VERBOSE: Device REMOTE sent an index update for "SyncBox" with 1 items
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now syncing
[LOCAL] 13:20:17 INFO: Folder "SyncBox" isn't making progress. Pausing puller for 1m0s.
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Summary for folder "SyncBox" is map[globalDirectories:10 localBytes:11787503 inSyncBytes:11787501 globalDeleted:32 needBytes:2 needFiles:1 inSyncFiles:53 needSymlinks:0 localSymlinks:0 needDirectories:0 sequence:801 needDeletes:0 localDirectories:10 localDeleted:32 version:801 globalBytes:11787503 localFiles:54 state:scanning globalFiles:54 globalSymlinks:0]
[LOCAL] 13:20:17 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 100%
[LOCAL] 13:20:17 VERBOSE: Local change detected in folder "": modified file C:\Documents\SyncBox\test\testfile.txt
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now scanning
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Device REMOTE sent an index update for "SyncBox" with 1 items
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now syncing
[LOCAL] 13:20:17 VERBOSE: Started syncing "SyncBox" / "test\\testfile.txt" (metadata file)
[LOCAL] 13:20:17 VERBOSE: Finished syncing "SyncBox" / "test\\testfile.txt" (metadata file): Success
[LOCAL] 13:20:17 VERBOSE: Remote change detected in folder "": modified file C:\Documents\SyncBox\test\testfile.txt
[LOCAL] 13:20:17 VERBOSE: Folder "SyncBox" is now idle
[LOCAL] 13:20:17 VERBOSE: Summary for folder "SyncBox" is map[needDeletes:0 globalFiles:54 globalSymlinks:0 inSyncBytes:11787503 needBytes:0 inSyncFiles:54 localDeleted:32 globalBytes:11787503 localSymlinks:0 sequence:804 needSymlinks:0 version:804 needDirectories:0 globalDeleted:32 needFiles:0 state:idle localFiles:54 localDirectories:10 localBytes:11787503 globalDirectories:10]
[LOCAL] 13:20:17 VERBOSE: Completion for folder "SyncBox" on device REMOTE is 100%
```

**Comment by calmh** [comment]
What happens without inotify stuff?

**Comment by canton7** [comment]
There's no inotify on either device (sorry, meant to say, then forgot).

**Comment by calmh** [comment]
Right, I can reproduce the multiple scanning/idle flips, seems to be caused by the scanning that happens when the local modification is detected during sync. Looks a bit odd and should probably be investigated but doesn't seem to cause any harm.

Conflict files being picked up on the next scan is expected behavior.

**Comment by calmh** [comment]
On 15 Feb 2017, at 15:22, Jakob Borg <jakob@nym.se> wrote:
> 
> Right, I can reproduce the multiple scanning/idle flips, seems to be caused by the scanning that happens when the local modification is detected during sync. Looks a bit odd and should probably be investigated but doesn't seem to cause any harm.

First guess as to the cause is probably the we trigger the scan and then immediately retry the pull, but the scan hasn't had time to be committed to the database yet, goto 10.

**Comment by canton7** [comment]
Yeah, things get in sync *eventually* so long as the devices are set to rescan every so often. If the rescan interval is set high, though, this could be a long time.

The bits that concern me are:

1. Multiple rescans, as you reproduced.
1. (Scenario 1): End up in "Out of Sync" for a minute in what should be a normal, simple situation
2. (Scenario 1): "Out of Sync Items" modal shows the same file 11 times, even though the link that was clicked to show the modal only talks about 1 item.
3. (Scenario 2): What happens doesn't make sense. Only Remote was rescanned - Local wasn't. I don't see how a conflict file could be created on Remote (since Remote shouldn't know that the file has changed on Local yet). This is probably just an expectation mismatch.

**Comment by calmh** [comment]
When syncing a file, we check that the file we are about to replace has the same metadata as the file in the index. If it does not, we trigger a scan of it and postpone the sync until it's been scanned. This is what sort of fails, and the puller gets paused for a minute, and you end up out of sync waiting for the next pull. That then happens a minute later, and then the conflict file is generated.

**Comment by imsodin** [comment]
Quite a lot of things about how scanning and pulling interacts changed since this ticket and I can't reproduce the scanning/idle flips nor the out-of-sync state (conflicts are still picked up only after the next scan, but that's expected and hardly noticeable now with watching for changes).
